pallene/typedecl.lua

```diff
@@ -82,4 +82,35 @@ function typedecl.declare(module, mod_name, type_name, constructors)
     end
 end
 
+-- Returns true if the 2nd argument is a prefix of the first.
+--
+-- @param tag: The type name (string)
+-- @param tag_prefix: The prefix to test (string)
+
+function typedecl.tag_matches(tag, tag_prefix)
+    return type(tag) == "string"
+        and string.find(tag, tag_prefix, 1, true) ~= nil
+end
+
+-- Return true if the argument tag is has "types.T" as prefix
+--
+-- @param tag Tag name
+
+function typedecl.tag_is_type(tag)
+    return typedecl.tag_matches(tag, "types.T")
+end
+
+-- Throw an error at the given type tag.
+--
+-- @param tag The type tag at which the error is to be thown (string)
+-- @param message The optional error message. (?string)
+
+function typedecl.tag_error(tag, message)
+    local error_msg = message
+        and string.format("error at tag '%s': %s", tag, message)
+        or string.format("error at tag '%s'.", tag)
```

Comment:
What do you think about changing the error message to not mention the word "tag". This way it would also be usable in other kinds of if-elseif like the binary operators and the builtin function names. Suggestion:

```
function typedecl.tag_error(tag, message)
    message = message or "input has the wrong type or an elseif case is missing"
    error(string.format("the if-elseif did not handle case %s: %s", tag, message))
end
```

Reply:
Yes, I think that would allow for more uses of `tag_error`.
My reasoning was that the function name implies the error is always associated with a tag.

However I do agree this allows for more liberal use of the function. I'll change it. :+1: